core/src/test/java/org/elasticsearch/cluster/serialization/ClusterSerializationTests.java

```diff
@@ -154,4 +154,58 @@ public void testSnapshotDeletionsInProgressSerialization() throws Exception {
         assertThat(stateAfterDiffs.custom(SnapshotDeletionsInProgress.TYPE), notNullValue());
     }
 
+    public void testObjectReuseWhenApplyingClusterStateDiff() throws Exception {
```

Comment:
while we're at it , can you check more than just the routing table and the index meta data? we have blocks, nodes and customs (out of the top of my head).